File: frontend/src/api/hodlcoin.ts
/**
 * HodlCoin Protocol API
 *
 * TypeScript types and invoke wrappers for Phoenix HodlCoin protocol Tauri commands.
 */

import { invoke } from '@tauri-apps/api/core'


// =============================================================================
// Type Definitions
// =============================================================================

export interface HodlBankState {
  bankBoxId: string
  singletonTokenId: string
  hodlTokenId: string
  hodlTokenName: string | null
  totalTokenSupply: number
  precisionFactor: number
  minBankValue: number
  devFeeNum: number
  bankFeeNum: number
  reserveNanoErg: number
  hodlTokensInBank: number
  circulatingSupply: number
  priceNanoPerHodl: number
  tvlNanoErg: number
  totalFeePct: number
  bankFeePct: number
  devFeePct: number
}

export interface HodlMintPreview {
  ergDeposited: number
  hodlTokensReceived: number
  pricePerToken: number
  minerFee: number
  totalErgCost: number
}

export interface HodlBurnPreview {
  hodlTokensSpent: number
  ergReceived: number
  bankFeeNano: number
  devFeeNano: number
  ergBeforeFees: number
  pricePerToken: number
  minerFee: number
}

// =============================================================================
// API Functions
// =============================================================================

export async function getHodlCoinBanks(): Promise<HodlBankState[]> {
  return await invoke<HodlBankState[]>('get_hodlcoin_banks')
}

export async function previewHodlCoinMint(
  singletonTokenId: string,
  ergAmount: number,
): Promise<HodlMintPreview> {
  return await invoke<HodlMintPreview>('preview_hodlcoin_mint', {
    singletonTokenId,
    ergAmount,
  })
}

export async function previewHodlCoinBurn(
  singletonTokenId: string,
  hodlAmount: number,
): Promise<HodlBurnPreview> {
  return await invoke<HodlBurnPreview>('preview_hodlcoin_burn', {
    singletonTokenId,
    hodlAmount,
  })
}

export async function buildHodlCoinMintTx(
  singletonTokenId: string,
  ergAmount: number,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('build_hodlcoin_mint_tx', {
    singletonTokenId,
    ergAmount,
    userUtxos,
    currentHeight,
  })
}

export async function buildHodlCoinBurnTx(
  singletonTokenId: string,
  hodlAmount: number,
  userUtxos: object[],
  currentHeight: number,
): Promise<object> {
  return await invoke<object>('build_hodlcoin_burn_tx', {
    singletonTokenId,
    hodlAmount,
    userUtxos,
    currentHeight,
  })
}

// =============================================================================
// Helper Functions
// =============================================================================

export function formatHodlPrice(priceNano: number): string {
  return (priceNano / 1_000_000_000).toLocaleString(undefined, {
    minimumFractionDigits: 6,
    maximumFractionDigits: 9,
  })
}

File: crates/citadel-api/src/services/hodlcoin.rs
//! HodlCoin use-case orchestration: bank discovery, preview math, tx building.

use ergo_node_client::NodeClient;

use super::error::{IntoServiceError, ServiceResult};
use crate::AppState;

async fn find_bank(
    client: &NodeClient,
    singleton_token_id: &str,
) -> ServiceResult<hodlcoin::HodlBankState> {
    let banks = hodlcoin::discover_banks(client).await.into_service()?;
    banks
        .into_iter()
        .find(|b| b.singleton_token_id == singleton_token_id)
        .ok_or_else(|| format!("Bank not found: {}", singleton_token_id))
}

pub async fn get_banks(state: &AppState) -> ServiceResult<Vec<hodlcoin::HodlBankState>> {
    let client = state.require_node_client().await?;
    hodlcoin::discover_banks(&client).await.into_service()
}

pub async fn preview_mint(
    state: &AppState,
    singleton_token_id: &str,
    erg_amount: i64,
) -> ServiceResult<hodlcoin::HodlMintPreview> {
    if erg_amount <= 0 {
        return Err("Amount must be greater than 0".to_string());
    }

    let client = state.require_node_client().await?;
    let bank = find_bank(&client, singleton_token_id).await?;

    let tokens_received = hodlcoin::mint_amount(
        bank.reserve_nano_erg,
        bank.circulating_supply,
        bank.precision_factor,
        erg_amount,
    );

    let miner_fee = citadel_core::constants::TX_FEE_NANO;
    let min_box = citadel_core::constants::MIN_BOX_VALUE_NANO;

    Ok(hodlcoin::HodlMintPreview {
        erg_deposited: erg_amount,
        hodl_tokens_received: tokens_received,
        price_per_token: bank.price_nano_per_hodl,
        miner_fee,
        total_erg_cost: erg_amount + miner_fee + min_box,
    })
}

pub async fn preview_burn(
    state: &AppState,
    singleton_token_id: &str,
    hodl_amount: i64,
) -> ServiceResult<hodlcoin::HodlBurnPreview> {
    if hodl_amount <= 0 {
        return Err("Amount must be greater than 0".to_string());
    }

    let client = state.require_node_client().await?;
    let bank = find_bank(&client, singleton_token_id).await?;

    let burn_result = hodlcoin::burn_amount(
        bank.reserve_nano_erg,
        bank.circulating_supply,
        bank.precision_factor,
        hodl_amount,
        bank.bank_fee_num,
        bank.dev_fee_num,
    );

    let miner_fee = citadel_core::constants::TX_FEE_NANO;

    Ok(hodlcoin::HodlBurnPreview {
        hodl_tokens_spent: hodl_amount,
        erg_received: burn_result.erg_to_user,
        bank_fee_nano: burn_result.bank_fee,
        dev_fee_nano: burn_result.dev_fee,
        erg_before_fees: burn_result.before_fees,
        price_per_token: bank.price_nano_per_hodl,
        miner_fee,
    })
}

pub async fn build_mint_tx(
    state: &AppState,
    singleton_token_id: &str,
    erg_amount: i64,
    user_utxos: Vec<ergo_tx::Eip12InputBox>,
    current_height: i32,
) -> ServiceResult<ergo_tx::Eip12UnsignedTx> {
    if erg_amount <= 0 {
        return Err("Amount must be greater than 0".to_string());
    }

    let client = state.require_node_client().await?;
    let bank = find_bank(&client, singleton_token_id).await?;

    let bank_box = client
        .get_eip12_box_by_id(&bank.bank_box_id)
        .await
        .map_err(|e| format!("Failed to fetch bank box: {}", e))?;

    let user_ergo_tree = user_utxos[0].ergo_tree.clone();

    hodlcoin::build_mint_tx_eip12(
        &bank_box,
        &bank,
        erg_amount,
        &user_utxos,
        &user_ergo_tree,
        current_height,
    )
    .into_service()
}

pub async fn build_burn_tx(
    state: &AppState,
    singleton_token_id: &str,
    hodl_amount: i64,
    user_utxos: Vec<ergo_tx::Eip12InputBox>,
    current_height: i32,
) -> ServiceResult<ergo_tx::Eip12UnsignedTx> {
    if hodl_amount <= 0 {
        return Err("Amount must be greater than 0".to_string());
    }

    let client = state.require_node_client().await?;
    let bank = find_bank(&client, singleton_token_id).await?;

    let bank_box = client
        .get_eip12_box_by_id(&bank.bank_box_id)
        .await
        .map_err(|e| format!("Failed to fetch bank box: {}", e))?;

    let user_ergo_tree = user_utxos[0].ergo_tree.clone();

    hodlcoin::build_burn_tx_eip12(
        &bank_box,
        &bank,
        hodl_amount,
        &user_utxos,
        &user_ergo_tree,
        current_height,
    )
    .into_service()
}
